File: lectures/week07-advanced-software-design/r-starter/demo_bug.R
# Run this to see the planted bug in action.
#
# Expected (correct) behavior: mean_by_site() returns a real number for
# every site, even when one of that site's lab values is missing (NA) --
# a missing value should be excluded from the mean, not silently wipe out
# the whole site's summary.
#
# What actually happens: source this file. The DC row's mean_value is NA,
# not the mean of the two valid values (4.2 and 3.8) -- one NA in the input
# silently erases the entire site's summary. No error, no warning.
#
# Debugging exercise (see ../practical.md): devtools::load_all(".") from
# this package's root, then set a breakpoint inside mean_by_site()
# (R/stats.R) -- RStudio's breakpoint gutter or browser() both work -- and
# step through the loop, inspecting site_rows$value for the DC group. Once
# you see what's happening, fix R/stats.R so a single NA doesn't erase the
# whole site's mean. Don't just drop NA rows here in this demo script
# instead.

# devtools::load_all(".")  # run from the r-starter/ package root first

labs <- data.frame(
  site = c("DC", "DC", "DC", "VA", "VA"),
  value = c(4.2, 3.8, NA, 5.0, 5.4)
)

print(mean_by_site(labs))
cat("Bug check: the DC row's mean_value should be a number, not NA.\n")

File: lectures/week07-advanced-software-design/python-starter/demo_bug.py
"""Run this to see the planted bug in action.

Expected (correct) behavior: calling summarize_by_site() with two different
batches of patient records should return two independent summaries -- the
second call's numbers should not be influenced by data passed to an earlier
call.

What actually happens: run this script and look at `second_summary`. It
still contains the "DC" site from the *first* call, and "VA"'s mean is
averaged across data from both calls, even though `second_batch` below has
only one VA record. No error, no warning -- just a wrong (contaminated)
answer, the same "confident but wrong" failure mode this course has flagged
since Week 4.

Debugging exercise (see ../practical.md): set a breakpoint inside
summarize_by_site() (in stats.py), step through both calls, and inspect
`_cache` before and after each call. Once you see what's happening, fix
stats.py so each call is independent -- don't just delete the docstring or
work around the symptom here in this script.
"""

from patientkit import summarize_by_site

first_batch = [
    {"site": "DC", "value": 4.2},
    {"site": "DC", "value": 3.8},
]
first_summary = summarize_by_site(first_batch)
print("First call:", first_summary)

second_batch = [
    {"site": "VA", "value": 5.0},
]
second_summary = summarize_by_site(second_batch)
print("Second call:", second_summary)
print("Bug check: 'DC' should NOT appear in the second call's output.")
